### PR TITLE
[NETBEANS-341] Added initial files required for Snap packaging

### PR DESCRIPTION
**Pull request by lkishalmi**
This files letnetbeans to be packaged with Snapraft for Linux distributions.
If you have snapcraft installed, just type snapcraft in the project root directory.

### CODE REVIEW

File: snap/gui/netbeans.desktop [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+[Desktop Entry]
+Type=Application
+Encoding=UTF-8
+Name=NetBeans Dev
+Comment=The Smater Way to Code
```
Comment (andre060989) [inline snap/gui/netbeans.desktop]: typo

**Comment by jlahoda** [comment]
I wonder: is there a chance to move this into some existing directory (or a sub-directory of thereof)? E.g. nbbuild/packaging/snap, or ide/packaging/snap something line that? I think that would be cleaner, and also would help to pack that directory into a source distro (because currently with this patch, I don't think this directory would be packed into the source zip).

File: snap/gui/netbeans.desktop (hunk 1 above)
Comment (lkishalmi) [reply snap/gui/netbeans.desktop]: Thanks!. I Try to move the snap dir around and fix this typo as well!

**Comment by lkishalmi** [comment]
Thanks for the comments! I moved the snap folder under nbbuild/packaging snapcraft shall be executed from that directory.

**Comment by geertjanw** [comment]
Is there anything preventing this from being merged?

**Comment by matthiasblaesing** [comment]
I tested the build on Ubuntu 17.10. I run `snapcraft` in the `nbbuild/packaging` directory. If I interpret the snapcraft.yml file correctly it expects `SNAP_ARCH` to be set and this is not the case:

```
matthias@athena:~/src/incubator-netbeans/nbbuild/packaging$ snapcraft 
Preparing to pull build 
Pulling build 
Preparing to build build 
Building build 
Error: JAVA_HOME is not defined correctly.
  We cannot execute /usr/lib/jvm/java-8-openjdk-/bin/java
Traceback (most recent call last):
  File "/usr/bin/snapcraft", line 11, in <module>
    load_entry_point('snapcraft==2.40+17.10', 'console_scripts', 'snapcraft')()
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 564, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 2662, in load_entry_point
    return ep.load()
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 2316, in load
```

The big question and relevant for the project: What is the result of the `snapcraft` invokation? If for example OpenJDK gets bundled, the result can't be distributed by the Apache Netbeans project (per my understanding), as the license (GPLv2-CP) is explicitly excluded from acceptable licenses for dependencies.

**Comment by lkishalmi** [comment]
Thank you for testing this PR!
I've tested on 18.04 Beta lately. Maybe SNAP_ARCH is a newly introduced provided environment variable. I have snapcraft installed as snap package:
snapcraft                2.40.1                   1345  stable    canonical  classic

I could make the package more strict to generate amd64 compatible package only. I think most developers are already on 64 bit platform.

As of dependencies. The snapcraft.yml just downloads the packages required to build NetBeans. Line 43 and 50 is responsible for the content of the package which indicates that only those files are going to be included which has been built by the Ant build system and placed into the netbeans directory.
Hence the resulting package would require classic confinement.

**Comment by lkishalmi** [comment]
@geertjanw On OpenJDK: AFAIK NB 9.x shall be built with Java 8. Unfortunately no legal source of debian package can be installed non-interactively of Oracle JDK, that's why the OpenJDK package is being used. It is not get into the distribution though.

**Comment by lkishalmi** [comment]
I've also rebased this branch from master, so It could be built on Ubuntu 18.04 as well which ships Ant 1.10. I've also tested the build on bare 16.04 in LXC container an that works as well.
There could be some improvement done regarding the details of the packaging regarding excluding some other platform binaries, but this should be a good start for providing snap packages from NetBeans,

**Comment by lkishalmi** [comment]
The snap package from this updated branch is available as:
```
wget https://s3-us-west-2.amazonaws.com/lkishalmi-us-west-2-public/snaps/netbeans_9.0-rc_amd64.snap
sudo snap install netbeans_9.0-rc_amd64.snap --dangerous --classic
```

**Comment by lkishalmi** [comment]
@matthiasblaesing Is there anything preventing this from being merged?

**Comment by matthiasblaesing** [comment]
@lkishalmi I was thinking that it would be nice if the version number could be determined from build properties, but I found, that the current version is not readily available.

I'll assume, that you'll take care, that the snap builds will continue to work.

In the long run an integration into the build system could be considered.

**Pull request merged**